pkg/kubelet/cm/deviceplugin/manager.go

```diff
@@ -610,58 +610,95 @@ func (m *ManagerImpl) allocateContainerResources(pod *v1.Pod, container *v1.Cont
 		if allocDevices == nil || len(allocDevices) <= 0 {
 			continue
 		}
-		startRPCTime := time.Now()
-		// devicePluginManager.Allocate involves RPC calls to device plugin, which
-		// could be heavy-weight. Therefore we want to perform this operation outside
-		// mutex lock. Note if Allocate call fails, we may leave container resources
-		// partially allocated for the failed container. We rely on updateAllocatedDevices()
-		// to garbage collect these resources later. Another side effect is that if
-		// we have X resource A and Y resource B in total, and two containers, container1
-		// and container2 both require X resource A and Y resource B. Both allocation
-		// requests may fail if we serve them in mixed order.
-		// TODO: may revisit this part later if we see inefficient resource allocation
-		// in real use as the result of this. Should also consider to parallize device
-		// plugin Allocate grpc calls if it becomes common that a container may require
-		// resources from multiple device plugins.
+		// Update internal cached podDevices state.
 		m.mutex.Lock()
-		e, ok := m.endpoints[resource]
+		m.podDevices.insert(podUID, contName, resource, allocDevices, nil)
 		m.mutex.Unlock()
-		if !ok {
-			m.mutex.Lock()
-			m.allocatedDevices = m.podDevices.devices()
+	}
+	return nil
+}
+
+// GetDeviceRunContainerOptions checks whether we have cached containerDevices
+// for the passed-in <pod, container> and returns its DeviceRunContainerOptions
+// for the found one. An empty struct is returned in case no cached state is found.
+func (m *ManagerImpl) GetDeviceRunContainerOptions(pod *v1.Pod, container *v1.Container) *DeviceRunContainerOptions {
+	podUID := string(pod.UID)
+	contName := container.Name
+	invokeRPCBoolFlag := false
+	for k, v := range container.Resources.Limits {
+		resource := string(k)
+		needed := int(v.Value())
+		glog.V(3).Infof("needs %d %s", needed, resource)
+		_, registeredResource := m.healthyDevices[resource]
+		_, allocatedResource := m.allocatedDevices[resource]
+		// Continues if this is neither an active device plugin resource nor
+		// a resource we have previously allocated.
+		if !registeredResource && !allocatedResource {
+			continue
+		}
+		m.mutex.Lock()
+		devices, resp := m.podDevices.containerDevices(podUID, contName, string(k))
+		if devices == nil {
+			glog.Errorf("No devices found allocated in local cache for pod %s, container %s, resource %s", podUID, contName, resource)
 			m.mutex.Unlock()
-			return fmt.Errorf("Unknown Device Plugin %s", resource)
+			return nil
 		}
+		if resp == nil {
```

Comment:
When is resp != nil ?

Reply:
Looking at the code it's probably on container restart.
If that's the case I expect container restart to issue an Allocate call (i.e: reset the device in the same state at every container start)

Reply:
Making Allocate call on each container restart have the side-effect that containers which were UP may not restart if device plugin has got failed.
How about having a configurable option, which will be configured by device plugin at registration time, `CacheUse`. Enabled by default, would mean the current PR functionality and if device plugin set it to `false`, `Allocate` will be invoked every time. WDYT?

Reply:
I guess the question you are asking really is what is the device plugin system expected to do.
The purpose of the device plugin is to setup a device before making it available in a container.

> Making Allocate call on each container restart have the side-effect that containers which were UP may not restart if device plugin has got failed.

Not making an Allocate call has also the side effect to break container reproducibility

> How about having a configurable option, which will be configured by device plugin at registration time, CacheUse. Enabled by default, would mean the current PR functionality and if device plugin set it to false, Allocate will be invoked every time. WDYT?

I agree with @lichuqiang do we have a use case for caching it ?
Is there a device plugin that needs caching ?

Reply:
@RenaudWasTaken hmm so you are advocating for making Allocate call for each container every time. I am fine with updating the changes if thats majority opinion. I personally would prefer to have it as a configuration option and using cache in default configuration. This would shorten container start-up time.

Reply:
> @RenaudWasTaken hmm so you are advocating for making Allocate call for each container every time. I am fine with updating the changes if thats majority opinion. 

If there is no use case I'm not sure it's worth.

> This would shorten container start-up time.

Let's do some actual metrics :)
It's also possible that checkpointing data is more expensive